hotham/src/systems/rendering.rs

```diff
@@ -25,72 +39,147 @@ pub fn rendering_system(
+                    },
+                    index_offset: primitive.index_buffer_offset,
+                    bounding_sphere: instance.bounding_sphere,
+                    visible: false,
                 });
             }
         }
     }
 
+    // This is the VERY LATEST we can possibly update our views, as the compute shader will need them.
     render_context.update_scene_data(views).unwrap();
+
+    // Execute the culling shader on the GPU.
     render_context.cull_objects(vulkan_context);
-    render_context.begin_pbr_render_pass(vulkan_context, swapchain_image_index);
 
-    let device = &vulkan_context.device;
-    let frame = &render_context.frames[render_context.frame_index];
-    let command_buffer = frame.command_buffer;
-    let draw_indirect_buffer = &frame.draw_indirect_buffer;
+    // Begin the render pass, bind descriptor sets.
+    render_context.begin_pbr_render_pass(vulkan_context, swapchain_image_index);
 
+    // Parse through the cull buffer and record commands. This is a bit complex.
     unsafe {
-        if USE_MULTI_DRAW_INDIRECT {
-            device.cmd_draw_indexed_indirect(
+        let device = &vulkan_context.device;
+        let frame = &mut render_context.frames[render_context.frame_index];
+        let command_buffer = frame.command_buffer;
+        let draw_data_buffer = &mut frame.draw_data_buffer;
+        draw_data_buffer.clear();
+
+        let mut instance_offset = 0;
+        let mut current_primitive_id = u32::MAX;
+        let mut instance_count = 0;
+        let cull_data = frame.primitive_cull_data_buffer.as_slice();
+
+        for cull_result in cull_data {
+            // If we haven't yet set our primitive ID, set it now.
+            if current_primitive_id == u32::MAX {
+                current_primitive_id = cull_result.index_offset;
+            }
+
+            // We're finished with this primitive. Record the command and increase our offset.
+            if cull_result.index_offset != current_primitive_id {
+                let primitive = &primitive_map.get(&current_primitive_id).unwrap().primitive;
+
+                // Don't record commands for primitives which have no instances, eg. have been culled.
+                if instance_count > 0 {
+                    device.cmd_draw_indexed(
+                        command_buffer,
+                        primitive.indices_count,
+                        instance_count,
+                        primitive.index_buffer_offset,
+                        primitive.vertex_buffer_offset as _,
+                        instance_offset,
+                    );
+                }
+
+                current_primitive_id = cull_result.index_offset;
+                instance_offset += instance_count;
+                instance_count = 0;
+            }
+
+            // If this primitive is visible, increase the instance count and record its draw data.
+            if cull_result.visible {
+                draw_data_buffer.push(&cull_result.draw_data);
```

Comment:
It is tempting to postpone creating `draw_data` until here. Do you think that would make any sense? I think it could make the `primitive_cull_data_buffer` a lot smaller.

Reply:
I tried it and it works but I don't know if it makes any significant difference, see here: https://github.com/leetvr/hotham/compare/kanerogers/issue284...lazy-draw-data

Reply:
See my comments above re: DrawData but yes I agree with you